remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
+            .context("Couldn't transition from Attesting to Attested state")?;
+        Ok((client_identity, attestation_engine))
+    }
+}
+
+impl AttestationEngine<Server, Attesting> {
+    /// Finishes the remote attestation protocol handshake and derives session keys for
+    /// encrypting/decrypting messages from the client.
+    ///
+    /// Transitions [`AttestationEngine`] state from [`Attesting`] to [`Attested`] state.
+    pub fn process_client_identity(
+        mut self,
+        client_identity: &ClientIdentity,
+    ) -> anyhow::Result<AttestationEngine<Server, Attested>> {
+        if self.behavior.contains_peer_attestation() {
+            // Verify client transcript signature.
+            // Transcript doesn't include transcript signature from the client identity message.
+            let mut client_identity_no_signature = client_identity.clone();
+            client_identity_no_signature.transcript_signature = vec![];
+            self.transcript
+                .append(&client_identity_no_signature)
+                .context("Couldn't append client identity to transcript")?;
+            let verifier = SignatureVerifier::new(&client_identity.signing_public_key);
+            verifier
+                .verify(
+                    &self.transcript.get_sha256(),
+                    &client_identity.transcript_signature,
+                )
+                .context("Couldn't verify client transcript")?;
+
+            // Verify client attestation info.
+            let client_attestation_info = &client_identity
+                .attestation_info
+                .as_ref()
+                .context("Couldn't get client attestation info")?;
+            let expected_tee_measurement = self
+                .behavior
+                .get_expected_tee_measurement()
+                .as_ref()
+                .context("Couldn't get expected TEE measurement")?;
+            verify_attestation_info(&client_attestation_info, &expected_tee_measurement)
+                .context("Couldn't verify client attestation info")?;
+        }
+
+        // Agree on session keys and create an encryptor.
+        let attestation_engine = self
+            .transition(&client_identity.ephemeral_public_key)
+            .context("Couldn't transition from Attesting to Attested state")?;
+        Ok(attestation_engine)
+    }
+}
+
+impl<P> AttestationEngine<P, Attested>
+where
+    P: AttestationParticipant,
+{
+    /// Returns an encryptor created based on the negotiated ephemeral keys.
+    pub fn get_encryptor(self) -> AeadEncryptor {
+        self.state.encryptor
     }
 }
 
-impl AttestationEngine<UnattestedSelf> {
-    pub fn create(tee_certificate: &[u8]) -> anyhow::Result<Self> {
-        let key_negotiator = KeyNegotiator::create().context("Couldn't create key negotiator")?;
+pub trait AttestationParticipant {}
+impl AttestationParticipant for Client {}
+impl AttestationParticipant for Server {}
```

Comment:
This seems... unusual :) still not sure exactly what it does in terms of type safety; is there anything that this design prevents at compile time instead of run time? I cannot thin of anything obvious

Reply:
As I said in the previous comment, these were added more for readability and because each protocol engine state is basically a new class (since it operates on different Protobuf messages).